qcfractal/storage_sockets/sqlalchemy_socket.py

```diff
@@ -1135,33 +1135,80 @@ def add_results(self, record_list: List[ResultRecord]):
             Dict with keys: data, meta
             Data is the ids of the inserted/updated/existing docs
         """
-
         meta = add_metadata_template()
 
-        result_ids = []
-        with self.session_scope() as session:
-            for result in record_list:
+        result_ids = ["placeholder"] * len(record_list)
+
+        results_list = []
+        existing_res = {}
+        new_record_idx, duplicates_idx = [], []
+        conds = []
+
+        conds = [
+            and_(
+                ResultORM.program == res.program,
+                ResultORM.driver == res.driver,
+                ResultORM.method == res.method,
+                ResultORM.basis == res.basis,
+                ResultORM.keywords == res.keywords,
+                ResultORM.molecule == res.molecule,
+            )
+            for res in results_list
+        ]
 
-                doc = session.query(ResultORM).filter_by(
-                    program=result.program,
-                    driver=result.driver,
-                    method=result.method,
-                    basis=result.basis,
-                    keywords=result.keywords,
-                    molecule=result.molecule,
+        with self.session_scope() as session:
+            docs = session.query(ResultORM).filter(or_(*conds)).all()
+
+            existing_res = {
+                (doc.program, doc.driver, doc.method, doc.basis, str(doc.keywords), str(doc.molecule)): doc
+                for doc in docs
+            }
+            for i, result in enumerate(record_list):
+                # necessary since result will be used to search through found items
+                if result.basis == "":
```

Comment:
This should not be need as the `basis` is correctly automatically coerced:

```
>>> ret1 = ResultRecord(driver='energy', program='a', molecule=5, method='test', basis='')
>>> type(ret1.basis)
<class 'NoneType'>
```

Reply:
Again, instantiating a ```ResultRecord``` object with the ```result.basis``` that was provided can be the way to go for comparison on the found items dictionary, but this one seems more efficient.

Reply:
The code requires a list of `ResultRecord` object which will have already correctly coerced the `result.basis`. This seems like an unnecessary `if` statement which is double true since the coercion logic is more complex than shown.

Reply:
moved to if the ```idx``` assignment part.